byterun/meta.c

```diff
@@ -49,40 +50,84 @@ CAMLprim value caml_get_section_table(value unit)
                                      caml_section_table_size);
 }
 
-CAMLprim value caml_reify_bytecode(value prog, value len)
+struct bytecode {
+  code_t prog;
+  asize_t len;
+};
+#define Bytecode_val(p) ((struct bytecode*)Data_abstract_val(p))
+
+CAMLprim value caml_reify_bytecode(value ls_prog, value debuginfo, value digest_opt)
 {
+  CAMLparam3(ls_prog, debuginfo, digest_opt);
+  CAMLlocal4(clos, bytecode, retval, s);
   struct code_fragment * cf = caml_stat_alloc(sizeof(struct code_fragment));
-  value clos;
+  code_t prog;
+  asize_t len, off;
+  int i;
 
+  /* Convert from LongString.t to contiguous buffer */
+  len = 0;
+  for (i = 0; i < Wosize_val(ls_prog); i++) {
+    s = Field(ls_prog, i);
+    len += caml_string_length(s);
+  }
+
+  prog = (code_t)caml_stat_alloc(len);
+  off = 0;
+  for (i = 0; i < Wosize_val(ls_prog); i++) {
+    size_t s_len;
+    s = Field(ls_prog, i);
+    s_len = caml_string_length(s);
+    memcpy((char*)prog + off, Bytes_val(s), s_len);
+    off += s_len;
+  }
+
+  caml_add_debug_info(prog, Val_long(len), debuginfo);
   cf->code_start = (char *) prog;
-  cf->code_end = (char *) prog + Long_val(len);
-  cf->digest_computed = 0;
+  cf->code_end = (char *) prog + len;
+  if (Is_block(digest_opt)) {
+    memcpy(cf->digest, String_val(Field(digest_opt, 0)), 16);
+    cf->digest_computed = 1;
+  } else {
+    cf->digest_computed = 0;
+  }
```

Comment:
I suppose this is handling a `string option` parameter? It would be nice if maybe the function could have a comment with the expected OCaml type of the `value` arguments passed.

Reply:
Now with match syntax!